037-pluggable-pod-security-profiles.md

```diff
@@ -0,0 +1,196 @@
+```
+
+While users can configure the security context, it is not user-friendly.
+The Pod security context does not cover all the different options required by the _restricted_ profile.
+So the security context needs to be basically configured for every container (including init containers, sidecars etc.).
+So just for a single Kafka cluster, it might need to be configured on 9 different places of the Kafka CR (ZooKeeper, Kafka broker, Kafka init-container, Topic Operator, User Operator, TLS sidecar, Kafka Exporter, Cruise Control, another TLS sidecar).
+
+## Proposal
+
+This proposal suggests to add a pluggable mechanism for users to provide their own plugins to configure the Pod and container security contexts of all Strimzi components.
+Out of the box, Strimzi will provide implementation supporting two profiles:
+* The _baseline_ provider which will mirror the current behavior and can be used when running under the _baseline_ Kubernetes profile
+* The _restricted_ provider which will allow to run under the _restricted_ Kubernetes profile
+
+The _baseline_ profile will be used by default.
+So users will not see any change by default.
+But if they decide to, they will be abe to switch to the _restricted_ profile.
+
+In addition, any user who will need some special configuration, will be able to provide a custom provider plugin.
+This will allow them to use their own configurations without specifying them in all custom resources in the `.template` sections.
+This might be useful for users who use some custom security tooling and might have their own requirement for the security configuration.
+But regular users will not need to write any custom plugins.
+
+The puggable mechanism should also make it easy for the Strimzi itself to add any additional implementations in the future.
+
+### Plugin implementation
+
+#### Interface
+
+A new interface `PodSecurityProvider` will be added to the `api` module.
+All provider plugins will implement this interface.
+The reason for having this interface in the `api` module is that this module is distributed though Maven repositories.
+Users will be able to write their implementations by using the `api` module as depednency.
+
+The interface will have methods for providing security contexts for all Pods created by Strimzi and all containers.
+The methods will take as a parameter the user-configured security context from the `.template` section of the custom resources.
+The methods for the pods using storage configuration (Kafka and ZooKeeper) will also have the storage configuration as parameter.
+These methods will have the default implementation which will always just return the user-specified context.
+For example:
+
+```java
+    default PodSecurityContext kafkaPodSecurityContext(PodSecurityContext userSecurityContext, Storage storage) {
+        return userSecurityContext;
+    }
+
+    default SecurityContext kafkaContainerSecurityContext(SecurityContext userSecurityContext, Storage storage) {
```

Comment:
Using the pattern of putting the storage in the signature like this would make for an awkward compatibility story should some pod which today don't use storage change in the future to require storage. 

Using an interface to represent the context for each method might be a more flexible pattern for coping with such changes. 

E.g.

```java
default PodSecurityContext kafkaPodSecurityContext(KafkaPodSecurityContext context) {
    return context.userSuppliedContext();
}

// where

interface KafkaPodSecurityContext {
    Storage storage();
    PodSecurityContext userSuppliedContext();
}

// and similar for all the other methods
```

Reply:
This is good idea. I added this to the proposal. However, the methods are in general the same. So I propose to use the same context for all of them.